From b7f7ebc3c639968fd0381a3e79cab361dd82a162 Mon Sep 17 00:00:00 2001
From: Xianzhong <xianzhong.li@nxp.com>
Date: Fri, 20 Jul 2018 22:11:11 +0800
Subject: [PATCH 4241/5242] MGS-3255-41 gpu: fix 6.2.4.p2 test_buffer failures

commit  bbcdb0d2964ad829823207a96e0d62c64eb3a10a from
https://source.codeaurora.org/external/imx/linux-imx.git

fix read buffer fence node for hw copy,
add cpu cache line check for write buffer,
invalidate cache after pipeline complete,
flush head and tail pages for user memory.

fix 630a849f3a4087a98b0425c261f474d178150307
 "MGS-3255-39 [#ccc] fix opencl cache issue"

Signed-off-by: Xianzhong <xianzhong.li@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../default/gc_hal_kernel_allocator_user_memory.c  |   15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/drivers/mxc/gpu-viv/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_user_memory.c b/drivers/mxc/gpu-viv/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_user_memory.c
index 4293ac9..8258cf8 100644
--- a/drivers/mxc/gpu-viv/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_user_memory.c
+++ b/drivers/mxc/gpu-viv/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_user_memory.c
@@ -201,6 +201,21 @@ static int import_page_map(struct um_desc *um,
         goto error;
     }
 
+    if (addr & ~PAGE_MASK)
+    {
+        dma_sync_single_for_device(galcore_device,
+                                   page_to_phys(pages[0]),
+                                   PAGE_SIZE,
+                                   DMA_TO_DEVICE);
+    }
+    if (page_count > 1 && ((addr + size) & ~PAGE_MASK))
+    {
+        dma_sync_single_for_device(galcore_device,
+                                   page_to_phys(pages[page_count-1]),
+                                   PAGE_SIZE,
+                                   DMA_TO_DEVICE);
+    }
+
     um->type = UM_PAGE_MAP;
     um->pages = pages;
 
-- 
1.7.9.5

